File: backend/src/AnimeShop/API/Stock/Items.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
module AnimeShop.API.Stock.Items (ItemsAPI, itemsServer) where

import AnimeShop.Prelude

type ItemsAPI =
  "items" :>
  (
    Get '[JSON] [Entity Item] :<|>
    ReqBody '[JSON] Item :> Post '[JSON] (Entity Item) :<|>
    ReqBody '[JSON] ItemId :> Delete '[JSON] () :<|>
    ReqBody '[JSON] (Entity Item) :> Patch '[JSON] ()
  )

itemsServer :: ServerT ItemsAPI App
itemsServer = getItems :<|> postItem :<|> deleteItem :<|> patchItem

getItems :: App [Entity Item]
getItems = runDb $ selectList [] []

postItem :: Item -> App (Entity Item)
postItem = runDb . insertEntity

deleteItem :: ItemId -> App ()
deleteItem = runDb . delete

patchItem :: Entity Item -> App ()
patchItem entity = runDb $ replace (entityKey entity) (entityVal entity)

File: frontend/src/reducers/stock/items.js
import { ADD_ITEM, GET_ITEMS, DELETE_ITEM } from '../../actions/types';

export function items(state = [], action) {
  switch (action.type) {
    case ADD_ITEM:
      return [ ...state, action.item];
    case GET_ITEMS:
      return action.items;
    case DELETE_ITEM:
      return state.filter(item => item.id !== action.id);
    default:
      return state;
  }
}
